File: app/src/main/java/com/example/simplehabitassignment/data/vos/models/CategoriesAndProgramsModelImpl.java
package com.example.simplehabitassignment.data.vos.models;


import com.example.simplehabitassignment.data.vos.CategoriesAndProgramsVO;
import com.example.simplehabitassignment.data.vos.ProgramsVO;
import com.example.simplehabitassignment.delegates.CategoriesAndProgramsResponseDelegate;
import com.example.simplehabitassignment.network.DataAgent;
import com.example.simplehabitassignment.network.RetrofitDA;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CategoriesAndProgramsModelImpl implements CategoriesAndProgramsModel {

    private static CategoriesAndProgramsModelImpl objInstance;
    private DataAgent dataAgent;
    private Map<String, ProgramsVO> programsMap;

    private CategoriesAndProgramsModelImpl() {
        dataAgent = RetrofitDA.getObjInstance();
        programsMap = new HashMap<>();
    }

    public static CategoriesAndProgramsModelImpl getObjInstance() {
        if (objInstance == null) {
            objInstance = new CategoriesAndProgramsModelImpl();
        }
        return objInstance;
    }

    @Override
    public void getCategoriesAndPrograms(String accessToken, int page, final CategoriesAndProgramsDelegate delegate) {

        dataAgent.getCategoriesAndPrograms(accessToken, page, new CategoriesAndProgramsResponseDelegate() {
            @Override
            public void onSuccess(List<CategoriesAndProgramsVO> categoriesProgramsList) {
                setProgramsMap(categoriesProgramsList);
                delegate.onSuccess(categoriesProgramsList);
            }

            @Override
            public void onFail(String errorMessage) {

                delegate.onFail(errorMessage);

            }
        });


    }

    private void setProgramsMap(List<CategoriesAndProgramsVO> categoriesProgramsList) {


        for (CategoriesAndProgramsVO c : categoriesProgramsList) {

            for (ProgramsVO p : c.getProgramsList()) {

                programsMap.put(p.getProgramId(), p);


            }
        }


    }


    public ProgramsVO getProgramById(String programId) {
        return programsMap.get(programId);
    }

}

File: app/src/main/java/com/example/simplehabitassignment/data/vos/models/CategoriesAndProgramsModelImpl.kt
package com.example.simplehabitassignment.data.vos.models


import com.example.simplehabitassignment.data.vos.CategoriesAndProgramsVO
import com.example.simplehabitassignment.data.vos.ProgramsVO
import com.example.simplehabitassignment.delegates.CategoriesAndProgramsResponseDelegate
import com.example.simplehabitassignment.network.DataAgent
import com.example.simplehabitassignment.network.RetrofitDA

import java.util.HashMap

class CategoriesAndProgramsModelImpl private constructor() : CategoriesAndProgramsModel {
    private val dataAgent: DataAgent
    private val programsMap: MutableMap<String, ProgramsVO>

    init {
        dataAgent = RetrofitDA.getObjInstance()
        programsMap = HashMap()
    }

    override fun getCategoriesAndPrograms(accessToken: String, page: Int, delegate: CategoriesAndProgramsModel.CategoriesAndProgramsDelegate) {

        dataAgent.getCategoriesAndPrograms(accessToken, page, object : CategoriesAndProgramsResponseDelegate {
            override fun onSuccess(categoriesProgramsList: List<CategoriesAndProgramsVO>) {
                setProgramsMap(categoriesProgramsList)
                delegate.onSuccess(categoriesProgramsList)
            }

            override fun onFail(errorMessage: String) {

                delegate.onFail(errorMessage)

            }
        })


    }

    private fun setProgramsMap(categoriesProgramsList: List<CategoriesAndProgramsVO>) {


        for (c in categoriesProgramsList) {

            for (p in c.programsList) {

                programsMap[p.programId] = p


            }
        }


    }


    fun getProgramById(programId: String): ProgramsVO? {
        return programsMap[programId]
    }

    companion object {

        private var objInstance: CategoriesAndProgramsModelImpl? = null

        fun getObjInstance(): CategoriesAndProgramsModelImpl {
            if (objInstance == null) {
                objInstance = CategoriesAndProgramsModelImpl()
            }
            return objInstance
        }
    }

}
